### PR TITLE
OpenSearch pagination

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR implements pagination for OpenSearch through the "Scroll" feature.
It also introduces GraphQL Connections (slightly modified to fit our needs) .

The new `pageSearchResultConnection` is meant to replace the `searchPages` query, moving some of the filtering logic to OpenSearch, increasing performance.

## ⚠️ Known issues

- The filtering that happens on the index does not allow for multiple account IDs, which means we cannot search for a user and its organization in one query.
- We are replacing the whole entity in the index every time an update happens, which means that historic entity versions are not searchable. This might not be a big deal, but if we move aggregations to the search index, this has to be thought through.

## 🔍 What does this change?

- Search loader now uses the HASH-API package for Database operations instead of custom queries for the sake of maintainability.
- Changes SearchLoader such that it pre-fetches grandparent pages for `Text` entities when adding to the `entities` OpenSearch index.
    - SearcLoader now indexes by `entityId` instead of `entityVersionId` such that "incomplete", historic text fields do not show up in search. Only the latest versions are present in the OpenSearch index, and we get natural index invalidation
- Adds a new GraphQL query `pageSearchResultConnection` that should deprecate `searchPages`
- Adds pagination to the search adapter and subsequently to the OpenSearch implementation of the search adapter
- Introduces a typed wrapper around OpenSearch search DSL, see `SearchField` in `packages/hash/backend-utils/src/search/adapter.ts`
- Changes to the `EntityType` model class, such that an EntityType can be fetched through its version ID (used by the Search Loader)

### 📜 Does this require a change to the docs?

- Perhaps we need new docs for pagination

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1201427269791756/f) (_internal_)
- [Connections proposal document](https://www.notion.so/hashintel/Cursor-based-pagination-proposal-0399616746944cd19309a660baf0ff5f) (_internal_)

## 🛡 What tests cover this?

- Unit tests for pagination

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2.  Sign in and create a new page under current user
3.  Type in a couple new text blocks with some shared word, for example: "text 1" "text 2" "the third text"
4. Run the following GQL queries
<details><summary>getAccountPages GQL</summary>

**Query**:

```graphql
query getAccountPages($accountId: ID!) {
  pageSearchResultConnection(
    accountId: $accountId
    query: "text"
    pageSize: 10
    # after: "" # to be used after first request
  ) {
    pageInfo {
      hasNextPage
      nextPageCursor
    }
    edges {
      node {
        score
        content
        text {
          entityId
        }
        page{
          entityId
        }
      }
    }
  }
}
```

**Data**:
```json
{
  "accountId": "..."
}
```
</details>

This should give back 2 of the 3 text entries, and a cursor token for the next page, which can be used in the "after"
## 📹 Demo
Running the test case given above yields the following JSON results:

**The first run**:
```json
{
  "data": {
    "pageSearchResultConnection": {
      "pageInfo": {
        "hasNextPage": true,
        "nextPageCursor": "eyJzZWVuQ291bnQiOjIsIm9wZW5TZWFyY2hDdXJzb3IiOiJGR2x1WTJ4MVpHVmZZMjl1ZEdWNGRGOTFkV2xrRFhGMVpYSjVRVzVrUm1WMFkyZ0JGbTgwU0d4cmEwZzBVbUpMYURKb1EyNHpabXhOVW5jQUFBQUFBQUFBSHhaR1NqSnBjMFF3ZUZKcFMyODBaRzE0VW5wbWRIZFIifQ=="
      },
      "edges": [
        {
          "node": {
            "score": 8.056613,
            "content": "text 1",
            "text": {
              "entityId": "68df6faa-dbe7-45bf-82ee-9f3ce4f0f92b"
            },
            "page": {
              "entityId": "7fb6efce-fc34-4a43-8ac7-39f3bd88f340"
            }
          }
        },
        {
          "node": {
            "score": 8.056613,
            "content": "text 2",
            "text": {
              "entityId": "3ea1cef4-66d9-4204-9ec9-fda9509a8404"
            },
            "page": {
              "entityId": "7fb6efce-fc34-4a43-8ac7-39f3bd88f340"
            }
          }
        }
      ]
    }
  }
}
```

**After using the resulting cursor**:
```json
{
  "data": {
    "pageSearchResultConnection": {
      "pageInfo": {
        "hasNextPage": false,
        "nextPageCursor": null
      },
      "edges": [
        {
          "node": {
            "score": 7.8048887,
            "content": "the third text ",
            "text": {
              "entityId": "d61cf7bc-713d-4ec9-920f-80849eddd32c"
            },
            "page": {
              "entityId": "7fb6efce-fc34-4a43-8ac7-39f3bd88f340"
            }
          }
        }
      ]
    }
  }
}
```
Notice that the `nextPageCursor` is null at this second page.

### CODE REVIEW

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:8]: I don't think this blocks merging, but would it be any trouble to add `pageCount` (or `numberOfPages`) as one of the `PageInfo` fields?

I appreciate that adding `hadPreviousPage` and `previousPageCursor` is not possible when using the opensearch implementation of pagination (based on internal prior discussions)

**Review by benwerner01 — APPROVED** [review]
Great work @thehabbos007 , this looks great and is working for me 🎉

Left a couple of questions/suggestions/minor thoughts, but spotted nothing that needs to be addressed before merging 👌

File: packages/hash/api/src/graphql/resolvers/pages/searchPages.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/pages/searchPages.ts:192]: Yeah this one was missing beforehand
https://github.com/hashintel/hash/blob/main/packages/hash/backend-utils/src/search/opensearch.ts#L137-L139
I'll try to see if i can find the/a relevant link

Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/pages/searchPages.ts:192]: Understood, sorry I didn't spot this wasn't introduced as part of this PR - I wouldn't invest too much time in finding that resource, maybe you can remove `For more see:` if you can't find it in a reasonable amount of time.

File: packages/hash/api/src/graphql/resolvers/pages/searchPages.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/pages/searchPages.ts:37]: Artifact of a different pagination implementation which used this method. Unexporting it :)

File: packages/hash/api/src/graphql/resolvers/pages/searchPages.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/pages/searchPages.ts:192]: I assume the intent was to reference https://opensearch.org/docs/latest/opensearch/query-dsl/full-text/#query-string
which seems highly relevant to the comment. Moved it elsewhere, though. This is not the place for it :)

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:41]: Would it be `(!query && !after)` instead? The suggested change would change semantics here if I'm not mistaken

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:49]: I do agree on this. I'll add a todo above the lines

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:41]: My mistake, I meant to suggest `(!query && !after)` here yes. The reason for this is because I wan't to avoid uses of `==` in a JS codebase where we primarily use `===`. I appreciate your usage here isn't wrong in any way, it's just harder to read what is really happening here in the context of codebase.

File: packages/hash/api/src/model/entityType.model.ts
Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:123]: We currently have that situation in search-loader. But it should be simple enough to change

Comment (thehabbos007) [reply packages/hash/api/src/model/entityType.model.ts:123]: So it seems that since we are subscribing on live changes from the database, the `entity_versions` table does not contain the `entityTypeId`, which is what we get through the realtime queue, so we'd have to look up the entity for the `entityVersionId` and get the `entityTypeId` from there. 

We are saving a database call for every modification on `entity_versions` if we don't look up the `entityTypeId`, which I would prefer. Thoughts on that?

Comment (benwerner01) [reply packages/hash/api/src/model/entityType.model.ts:123]: Hmm I see, I haven't considered that. In that case I think it's fine to leave it as is for now then, my instinct is that we'll want to modify how the search-loader receives updates - and not have this happen at the postgres level. But this is out of scope for this PR.

File: packages/hash/backend-utils/src/search/doc-types.ts
Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/doc-types.ts:17]: This is to comply with the GQL `EntityRef` type, makes conversion from the search index to a `PageSearchResult` trivial

Comment (benwerner01) [reply packages/hash/backend-utils/src/search/doc-types.ts:17]: Because as part of this PR we are only indexing the latest version of entities, perhaps instead of `EntityRef` we should be using `LatestEntityRef` in the GQL definitions for search-related things? This should let us remove `entityVersionId` here without sacrificing simplicity in that conversion process.

Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/doc-types.ts:17]: Hmm, I was not aware of `LatestEntityRef`. 
This would also touch on the existing `searchPages` resolver as the search result is a shared type. 
Would this influence anything on the frontend search functionality?

Comment (benwerner01) [reply packages/hash/backend-utils/src/search/doc-types.ts:17]: I did not implement FE search, but my guess would be that this shouldn't impact the current implementation of search (as it shouldn't really do anything with the `entityVersionId`s as far as I'm aware). I think this is worth looking into, as I'd like us to avoid passing around `entityVersionId` in places where it has no meaning.

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:16]: Sounds good, thank you!

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:114]: If this doesn't interfere with existing FE logic, I'm happy for this to be addressed in a follow-up (would be good to create a task for this though and link this thread)

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/opensearch.ts:258]: Good catch, thank you. Added a method to clear the cursor and use this both places (if statement above and this branch of the ternary)

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:8]: Adding this after ironing out comments above :)

File: packages/hash/backend-utils/src/search/doc-types.ts
Comment (benwerner01) [reply packages/hash/backend-utils/src/search/doc-types.ts:22]: Understood, in that case lets create a follow-up task 👍

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts]: Captures in [this task](https://app.asana.com/0/1200211978612931/1201892371894926/f) (_internal))

File: packages/hash/backend-utils/src/search/doc-types.ts
Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/doc-types.ts:22]: Captures in [this task](https://app.asana.com/0/1200211978612931/1201892371894926/f) (_internal_)

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts]: Captured in [this task](https://app.asana.com/0/1200211978612931/1201892371894932/f) (_internal_)

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (benwerner01) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:48]: Thanks, in that case I've added "Create `Text` entity model class" as a sub-task. Btw I noticed that linking to GH threads from asana is not going to work for this PR, as all comment threads won't be loaded by the browser when the page loads, so GH won't scroll to the relevant thread. Therefore I think it's worth being a bit more descriptive in the tasks.

File: packages/hash/backend-utils/src/search/doc-types.ts
Comment (benwerner01) [reply packages/hash/backend-utils/src/search/doc-types.ts:22]: Thank you :)

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:130]: Nit pick: `.sort` does not need to happen on a separate line, and we can destructure the args passed to the mapping function (instead of using `it`, which is a typo for `hit`?):
```suggestion
  const textMatches = hits
    .filter(
      (hit): hit is TextSearchHit =>
        hit.document.entityTypeId === textType.entityId,
    )
    .map<PageSearchResult>(({ score, document }) => ({
      score,
      page: document.belongsToPage!,
      /**
       * @todo: Currently we are not getting the (parent) block of a text system type in the search-loader
       * This means that scrolling to a block after selecting a search hit is not possible even though
       * the API implies it is. We will need to implement indexing of the "parent block" within the search-loader.
       */
      block: undefined,
      text: {
        accountId: document.accountId,
        entityId: document.entityId,
        entityVersionId: document.entityTypeVersionId,
      },
      content: document.fullTextSearch || "",
    }))
    .sort((matchA, matchB) => matchB.score - matchA.score);
```

File: packages/hash/search-loader/src/searchLoader.ts
Comment (benwerner01) [inline packages/hash/search-loader/src/searchLoader.ts:173]: **Blocking: ** this is not a robust way of determining whether a page is a page entity on the BE, as users could fool this by creating an entity type with the name `"Page"` in their namespace.

We should compare the IDs of the entity types instead (using `(await Page.getEntityType(...)).entityId` for instance)

**Review by benwerner01 — CHANGES_REQUESTED** [review]
Thanks for addressing my suggestions 🙂

Spotted one potential issue in one of the recent changes that was made, related to how we can check whether something is a system type on the BE. As it has the potential for introducing bugs for us down the line I think it's worth addressing as part of this PR. The rest looks good to me 👍

File: packages/hash/search-loader/src/searchLoader.ts
Comment (thehabbos007) [reply packages/hash/search-loader/src/searchLoader.ts:173]: We could load the page system type once when we `start` the search loader. What do you think about that instead of fetching it every time?

File: packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/paginationConnection/pageSearchResultConnection.ts:130]: Yes, good point. I left it separate as it is a mutating operation, but I don't have strong feelings about having it be a statement for itself :)

File: packages/hash/search-loader/src/searchLoader.ts
Comment (benwerner01) [reply packages/hash/search-loader/src/searchLoader.ts]: We can do that yes, you're right this will speed things up now although I'm not sure it's really worth the effort as long-term we'll definitely want some caching solution for _all_ system types. In general I've not worried about calls to `Page.getEntityType` for that reason.

Comment (thehabbos007) [reply packages/hash/search-loader/src/searchLoader.ts]: Yes good point. That has been mentioned before. I don't think it's an issue to do this basic caching here right now. 

I can make the call directly to the model class system type method if you prefer, though.

Comment (benwerner01) [reply packages/hash/search-loader/src/searchLoader.ts]: I'm happy for us to merge the temporary caching solution you've implemented, just want to make sure that we're on the same page that this should be removed once we have a generalised caching solution in place.

Comment (thehabbos007) [reply packages/hash/search-loader/src/searchLoader.ts]: Yes, and I appreciate that moving to model class methods will help us enable generalized caching too. It makes a lot of sense :)

**Review by benwerner01 — DISMISSED** [review]
Looks good to me, great work @thehabbos007  👍

**Comment by thehabbos007** [comment]
Thank you @benwerner01. Leaving this to be merged tomorrow if @CiaranMn has any comments :)

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (CiaranMn) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:5]: We would also ideally describe the purpose/summary of each type - once we have hash.dev up we can generate docs for the API directly from these files to show there.
```suggestion
  Information returned with paginated queries.
  
  Each pagination specifies if there are more results to request.
```

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (CiaranMn) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:10]:
```suggestion
    # kept in to comply with the Connections spec. 
    # @see https://relay.dev/graphql/connections.htm
```

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (CiaranMn) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:25]: I appreciate this is somewhat redundant description but it'll look nicer in auto-generated docs
```suggestion
  """
  An edge of a paginated query result.
  
  There are no cursors per-edge currently.
  """
```

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (CiaranMn) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:26]: We might want to make this a more generic `SearchResult` edge when introducing more OpenSearch-using queries - but let's worry about that then.

File: packages/hash/backend-utils/src/search/adapter.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/adapter.ts:102]:
```suggestion
   * @param params.cursor the search cursor to fetch hits from
```

File: packages/hash/backend-utils/src/search/doc-types.ts
Comment (CiaranMn) [reply packages/hash/backend-utils/src/search/doc-types.ts:22]: To note that the frontend isn't using search at the moment - it's been broken since the links refactor. 

Agree this is good to do in follow up, but when doing follow ups I wouldn't worry about breaking the frontend search.

File: packages/hash/backend-utils/src/search/opensearch-pagination.test.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch-pagination.test.ts:24]: hey, a generator, nice :)

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch.ts:54]: checking it's not `"AUTO"` is technically equivalent to checking if it's a number since we rely on the typing, but more quickly/easily conveys all permitted values.
```suggestion
      ([_, it]) =>
        it.fuzziness !== "AUTO" &&
        (!Number.isInteger(it.fuzziness) || it.fuzziness < 0),
```

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch.ts:239]: extreme nit
```suggestion
   * @param params.hitCount number of hits in the result
   * @param params.seenCount number of hits seen so far (not including the newly seen)
   * @param params.total number of hits for the whole search
   * @param params.openSearchCursor cursor used to receive the current results
   * @param params.nextOpenSearchCursor cursor to use to retrieve results after the current results
```

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch.ts:269]: What's an improper next cursor?

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch.ts:280]:
```suggestion
    // If there are not hits in the page, the newCursor will be undefined as well.
```

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [inline packages/hash/backend-utils/src/search/opensearch.ts:289]: If I've understood this correctly, this is a pretty serious limitation which we should discuss the consequences of and alternatives to - it'll make harder for API consumers to programmatically do work through pages of results, for example, if the work takes longer than 10 minutes. And prevent browser users from paging through results slowly (e.g. if they switch task for 10 minutes and then come back and hit the next page button).

I don't think we've spoken about `search_after` so it'd be good to understand the trade-offs involved.

**Review by CiaranMn — APPROVED** [review]
Thanks @thehabbos007 - great work.

I have a bunch of nits, mostly on comments, which you can take or leave.

I have a pretty big concern about the 10 minute cursor life, which we should discuss for follow up work. I'm assuming we can swap out the cursor implementation without affecting the API interface so I'd get this in now and we can figure out if we want to replace the internals later.

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/opensearch.ts:269]: One that is either closed or not wrapped with the extension type we have around the raw OpenSearch cursors

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/opensearch.ts:289]: I believe `search_after` requires us to have sortable ids (I added a section to the cursor document on notion to address the potential before starting work on this).

It would require we do some of the paging logic ourselves, which shouldn't be too bad, and we wouldn't have to have a close-timer on pagination. It would definitely be beneficial to use `search_after` for our purposes instead of scroll, though.

That being said, we can increase this timeout if we want to. I am not sure what the limitations on AWS hosted OpenSearch is in regards to this.

Comment (CiaranMn) [reply packages/hash/backend-utils/src/search/opensearch.ts:294]: Cool, thank you. Presumably it would only require us to have sortable ids if we want to return a sort that is sorted by ids - which I don't think we ever want to do. We could use `createdAt` as a default sort instead, if the user doesn't specify any. 

Let's talk more in the next refinement session about this.

Comment (thehabbos007) [reply packages/hash/backend-utils/src/search/opensearch.ts:294]: By the way, using `search_after` would allow us to do pagination in both directions by changing sort ordering

**Review by CiaranMn — APPROVED** [review]
Excellent!

File: packages/hash/backend-utils/src/search/opensearch.ts
Comment (CiaranMn) [reply packages/hash/backend-utils/src/search/opensearch.ts:294]: Even better!

**Pull request merged**
